### DISCUSSION TITLE
Time Out Error while trying to Update Server Value -under  Data Change Notification

### DISCUSSION THREAD

**Discussion by saimanoj567** [FreeOpcUa/python-opcua#1190] [Q&A]
Hello ,

I'm a student and New to OPC UA ( Pub- Sub ). Just Playing around in these concepts.. What woud be the possible ways to handle this scenario... 

```
from opcua import Client
from opcua import ua
import time

url = "opc.tcp://192.123.213.44:4840"
client= Client(url)
client.connect()
print("Client Connected")


class SubHandler(object):

    def datachange_notification(self, node, val, data):
        print(node,val)
        if val == True:   
            print("Starting analysis part")# when analysis status is true
            #some analysis script has to be  executed 
            #...........
            #........
            #updating the status as finished once the analyis part is done 
            change_analysis_status = ua.DataValue(ua.Variant(False, analysis_status.get_data_type_as_variant_type()))
            analysis_status.set_value(change_analysis_status)
            
analysis_status = client.get_node("ns=4;s=hello.analysisStart") #boolean

handler = SubHandler()
sub = client.create_subscription(500, handler)
handle1= sub.subscribe_data_change(analysis_status)
```

**Comment by AndreasHeine** [comment]
FYI:
Pub-Sub is not implemented! (It would use UDP Multicast, AMQP and MQTT)
https://reference.opcfoundation.org/v104/Core/docs/Part14/

the subscription which you are using is part of the "Subscription Service Set" (which uses the normal TCP based OPC Binary Protocol)!
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.13.1/

typically in the subhandler callback you should not do intense work like analysis or heavy calculations!!! (you are blocking the mainthread thats where the timeout came from)


if you want to process data from clientside you should look deeper into OPC UA Method:
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.11.1/

**Reply by saimanoj567** [reply]
@AndreasHeine . Thanks Andreas... I am Getting Time out error only when I try to update the Analysis ( boolean ) Tag value inside the call back. If remove the `change_analysis_status = ua.DataValue(ua.Variant(False, analysis_status.get_data_type_as_variant_type()))
            analysis_status.set_value(change_analysis_status)` then I dont get any TIME OUT ERROR ..

**Reply by AndreasHeine** [reply]
could you post the stacktrace?

**Reply by saimanoj567** [reply]
`Starting analysis part
61181700.12  4080707.52
Exception calling data change handler
Traceback (most recent call last):
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\subscription.py", line 153, in _call_datachange
    self._handler.datachange_notification(data.node, item.Value.Value.Value, event_data)
  File "opc_ua_sub_Try.py", line 29, in datachange_notification
    change_analysis_status = ua.DataValue(ua.Variant(False, analysis_status.get_data_type_as_variant_type()))
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\node.py", line 101, in get_data_type_as_variant_type
    result = self.get_attribute(ua.AttributeIds.DataType)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\node.py", line 275, in get_attribute
    result = self.server.read(params)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\client\ua_client.py", line 347, in read
    data = self._uasocket.send_request(request)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\client\ua_client.py", line 83, in send_request
    data = future.result(self.timeout)
  File "C:\Util\Anaconda3\lib\concurrent\futures\_base.py", line 437, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError
`

**Reply by oroulet** [reply]
yes you get a lock so you get a timeout somewhere. So do not do that. In fact if using opcua-asyncio that kind of things might works as long as you call await inside the callback

**Comment by oroulet** [comment]
I think there is a big warning in code saying something lik "do not try to call any us method INSIDE the datachange_notification callback. Everything will get locked"
I have yet to hear about a good reason to do that so most probably you have some architecture error in your program

**Reply by swamper123** [reply]
Quote from examples:

> 
> class SubHandler(object):
> 
>     """
>     Subscription Handler. To receive events from server for a subscription
>     data_change and event methods are called directly from receiving thread.
>     Do not do expensive, slow or network operation there. Create another
>     thread if you need to do such a thing
>     """

And from client.py:

> 
>    def create_subscription(self, period, handler):
>         """
>         Create a subscription.
>         returns a Subscription object which allow
>         to subscribe to events or data on server
>         handler argument is a class with data_change and/or event methods.
>         period argument is either a publishing interval in milliseconds or a
>         CreateSubscriptionParameters instance. The second option should be used,
>         if the opcua-server has problems with the default options.
>         These methods will be called when notfication from server are received.
>         See example-client.py.
>         Do not do expensive/slow or network operation from these methods
>         since they are called directly from receiving thread. This is a design choice,
>         start another thread if you need to do such a thing.
>         """

**Comment by saimanoj567** [comment]
how can we Set a value within datachange _notification (Method) OPC UA -SUB ? - Timeout Error with in methods ... I hope now its clear ?

```
from opcua import Client
from opcua import ua
import time

class SubHandler(object):

    def datachange_notification(self, node, val, data):
        print(node,val)
        if val == True:   
                change_machine_status = ua.DataValue(ua.Variant(False, analysis_status.get_data_type_as_variant_type()))
                machine_status.set_value(change_machine_status)

if __name__ == '__main__':

            url = "opc.tcp://192.168.178.59:4840"
            client= Client(url)
            client.connect()
            print("Client Connected")

            machine_status = client.get_node("ns=4;s=Hello.MachineStart") 
            handler = SubHandler()
            sub = client.create_subscription(400, handler)
            handle1= sub.subscribe_data_change(machine_status)

```
@AndreasHeine @oroulet  @swamper123

**Reply by saimanoj567** [reply]
The same Error occurs for this scenario ...

```
Exception calling data change handler
Traceback (most recent call last):
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\subscription.py", line 153, in _call_datachange
    self._handler.datachange_notification(data.node, item.Value.Value.Value, event_data)
  File "opc_ua_sub_Try.py", line 29, in datachange_notification
    change_analysis_status = ua.DataValue(ua.Variant(False, analysis_status.get_data_type_as_variant_type()))
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\node.py", line 101, in get_data_type_as_variant_type
    result = self.get_attribute(ua.AttributeIds.DataType)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\common\node.py", line 275, in get_attribute
    result = self.server.read(params)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\client\ua_client.py", line 347, in read
    data = self._uasocket.send_request(request)
  File "C:\Util\Anaconda3\lib\site-packages\opcua\client\ua_client.py", line 83, in send_request
    data = future.result(self.timeout)
  File "C:\Util\Anaconda3\lib\concurrent\futures\_base.py", line 437, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError
```

@AndreasHeine

**Reply by AndreasHeine** [reply]
i am not sure if its possible at all if the datachange callback is called the main thred is blocked the client cant do anything till the callback returns and in this locked situation you tell the client (which is blocked) write the value

what i would do is i would use the datachange callback to put it onto a queue an then i would process it outside the callback maybe in a different thread (make sure you use a threadlock for the queue and only put or take one item on or from the queue and release the lock after doing so)!

but still i am not sure why someone use a subscription for updating a variable.... opc ua uses client/server request/response it looks like you tr to invert it!? do you update the status on the same server you get the notification from? 

it look like you make recursion:
status change notification -> write the value (this triggers a status change notification) -> status change notification -> write the value (this triggers a status change notification)  ->status change notification -> write the value (this triggers a status change notification)  ->...

**Reply by saimanoj567** [reply]
yeah I update the status on the same server that I get the notification from.  My Goal is to perform some analysis based on tag value and update the value back into server to carry out next actions  --- Trying out the **robust way** .. **Focus : Latency**

**Reply by AndreasHeine** [reply]
tbh this is a terrible idea! atm you make a endless loop... 
i dont want to be rude!

 you should use a different programstructure

**Comment by starturtle** [comment]
Like mentioned recently: can you use `concurrent.futures`, event queue or similar to instead post a trigger to a different thread that will then perform the analysis as soon as the callback has completed and permits a context switch?

**Reply by saimanoj567** [reply]
Can I try this scenario with asynico ? Whether it would be a better choice ?

**Reply by swamper123** [reply]
I would recommened you, before you start with asyncio in python-opcua, to have a look on our project: [opcua-asyncio](https://github.com/FreeOpcUa/opcua-asyncio) , which is the sequal of python-opcua (this repo is in maintainance mode!).
The handling of an eventloop is, in my opinion, better and easier, than to multiple threads.

**Comment by oroulet** [comment]
@saimanoj567 learn threading and/or Future and you can do it. No need to discuss that stuff with 100 messages
